File: 20-valid-parentheses/20-valid-parentheses.java
import java.util.*;
class Solution {
    public boolean isValid(String s) {
         Stack<Character> st = new Stack<Character>();
        HashMap<Character,Character> mapping = new HashMap<Character,Character>();
        mapping.put('(',')');
        mapping.put('{','}');
        mapping.put('[',']');
        // while string has characters
        try{
        // keep pushing all opening brackets
            for(int x=0;x<s.length();x++){
            char i = s.charAt(x);
            if(mapping.containsKey(i)){
                st.push(i);
            }
            else{
                char topElement = st.pop();
                if(mapping.get(topElement)!=i){
                    return false;
                }
            }
        }
            // keep doing this

            // in the end stack should become empty
            
            // otherwise give false
            return st.isEmpty();
        }
        catch(Exception e){
            return false;
        }
    }
}

File: 20-valid-parentheses/20-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        st = deque()
        pair = {"]":"[","}":"{",")":"("}
        for c in s:
            if c in "[({":
                st.append(c)
            else:
                if len(st)==0 or st[-1]!=pair[c]:
                    return False
                st.pop()
        
        return len(st)==0